File: ios/ParkingKok/Detection/Trace/TraceEvent.swift
import Foundation

/// The wire vocabulary of a recorded trace
/// (docs/05_CROSS_PLATFORM_DOMAIN_CONTRACT.md §9, reusing §2's normalized events and §8's
/// fixture vocabulary verbatim).
///
/// Raw strings are the contract, so they are written out rather than derived from the
/// case names. A platform SDK enum must never reach this type: §9 forbids it, and a trace
/// whose vocabulary drifts from the fixture vocabulary cannot be converted into one.
enum TraceEventType: String, Sendable, Equatable, Codable, CaseIterable {
    case vehicleEnter = "vehicle_enter"
    case vehicleExit = "vehicle_exit"
    case walkingEnter = "walking_enter"
    case stationaryEnter = "stationary_enter"
    case stationaryExit = "stationary_exit"
    case location
    case locationQualityDegraded = "location_quality_degraded"
}

/// Coarse horizontal-accuracy band, the only form in which location quality is recorded.
///
/// **The thresholds are literals on purpose, and must stay that way.** They look like
/// `ReliableLocationPolicy.maximumHorizontalAccuracy` by coincidence only. That value is
/// a detection threshold destined for remote tuning
/// (docs/03_SYSTEM_ARCHITECTURE.md §10 `detector.reliableAccuracyMeters`); binding the
/// buckets to it would retroactively change what an already-recorded trace *means* the
/// moment the threshold moved, and a recording format has to stay comparable over time.
///
/// There is deliberately no bucket for a negative accuracy. docs/05 §5 calls such a fix
/// invalid and `LocationQualitySample.isValid` / `LocationFix.isValid` reject it at the
/// adapter boundary; one reaching the recorder is an adapter defect, and giving it a
/// bucket would swallow the defect instead of leaving it visible as a `location` event
/// with a negative accuracy and no quality transition.
enum LocationAccuracyBucket: String, Sendable, Equatable, Codable, CaseIterable, Comparable {
    case good
    case fair
    case poor

    static let goodMaximumMeters: Double = 20
    static let fairMaximumMeters: Double = 35

    /// `nil` for an invalid (negative) accuracy — see the type's note.
    init?(horizontalAccuracy: Double) {
        guard horizontalAccuracy >= 0 else { return nil }
        switch horizontalAccuracy {
        case ...Self.goodMaximumMeters: self = .good
        case ...Self.fairMaximumMeters: self = .fair
        default: self = .poor
        }
    }

    private var rank: Int {
        switch self {
        case .good: 0
        case .fair: 1
        case .poor: 2
        }
    }

    /// Ordered best-first, so `>` reads as "worse quality than".
    static func < (lhs: LocationAccuracyBucket, rhs: LocationAccuracyBucket) -> Bool {
        lhs.rank < rhs.rank
    }
}

/// One recorded observation, in the exact shape docs/05 §9 fixed.
///
/// **A flat struct with optional fields rather than an enum with associated values.** The
/// contract is a JSON object per event whose keys depend on `type`; Swift's synthesized
/// encoder already omits `nil` optionals, so this shape round-trips the contract without
/// a hand-written coder to keep in sync with it.
///
/// **There is no coordinate member, and adding one is never the fix for anything.** §9
/// bans coordinates from the trace because the file's whole purpose is to be copied off
/// the device; a latitude here turns every recorded trip into a parking-location log.
/// `distanceFromPreviousM` is the derived value that survives instead — see
/// `TraceRecorder`, which holds the previous fix in memory only and stores the metres.
struct TraceEvent: Sendable, Equatable, Codable {
    let type: TraceEventType
    /// Absolute epoch milliseconds. Conversion to a fixture's relative `t` happens in the
    /// converter, not here (§9 "변환").
    let atMillis: Int64
    let confidence: MotionConfidence?
    /// Metres.
    let accuracy: Double?
    /// Metres per second.
    let speed: Double?
    /// Metres travelled since the previous recorded `location` event, summed over the
    /// fixes that were downsampled away. Never a position.
    let distanceFromPreviousM: Double?
    let fromBucket: LocationAccuracyBucket?
    let toBucket: LocationAccuracyBucket?

    private init(
        type: TraceEventType,
        at date: Date,
        confidence: MotionConfidence? = nil,
        accuracy: Double? = nil,
        speed: Double? = nil,
        distanceFromPreviousM: Double? = nil,
        fromBucket: LocationAccuracyBucket? = nil,
        toBucket: LocationAccuracyBucket? = nil
    ) {
        self.type = type
        atMillis = date.traceMillis
        self.confidence = confidence
        self.accuracy = accuracy
        self.speed = speed
        self.distanceFromPreviousM = distanceFromPreviousM
        self.fromBucket = fromBucket
        self.toBucket = toBucket
    }

    static func motion(_ type: TraceEventType, at date: Date, confidence: MotionConfidence?) -> TraceEvent {
        TraceEvent(type: type, at: date, confidence: confidence)
    }

    static func location(
        at date: Date,
        accuracy: Double,
        speed: Double? = nil,
        distanceFromPreviousM: Double? = nil
    ) -> TraceEvent {
        TraceEvent(
            type: .location,
            at: date,
            accuracy: accuracy,
            speed: speed,
            distanceFromPreviousM: distanceFromPreviousM
        )
    }

    static func qualityDegraded(
        at date: Date,
        from: LocationAccuracyBucket,
        to: LocationAccuracyBucket
    ) -> TraceEvent {
        TraceEvent(type: .locationQualityDegraded, at: date, fromBucket: from, toBucket: to)
    }

    var date: Date {
        Date(traceMillis: atMillis)
    }
}

/// §9 records absolute time as epoch milliseconds, not ISO-8601 — the same unit Android
/// gets from `System.currentTimeMillis()`, so both platforms write the same number.
extension Date {
    var traceMillis: Int64 {
        Int64((timeIntervalSince1970 * 1000).rounded())
    }

    init(traceMillis: Int64) {
        self.init(timeIntervalSince1970: Double(traceMillis) / 1000)
    }
}

File: android/app/src/main/java/kr/parkingpin/app/domain/trace/TraceEvent.kt
package kr.parkingpin.app.domain.trace

import kr.parkingpin.app.domain.detection.MotionEventKind
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * One recorded event inside a [TraceSession], exactly as
 * docs/05_CROSS_PLATFORM_DOMAIN_CONTRACT.md §9 defines it.
 *
 * **No coordinate is representable here.** The schema's whole point is that a trace file
 * is copied off the device, so §9 restricts it to `type`/`atMillis`/`accuracy`/`speed`/
 * `distanceFromPreviousM`/`confidence`/buckets. There is no latitude or longitude field to
 * populate, by construction — the same structural choice
 * [kr.parkingpin.app.domain.location.LocationQualitySample] makes on the detection path.
 * `TraceSessionSerializationTest` enforces it against the encoded bytes.
 *
 * Flat rather than a sealed hierarchy: §9 specifies one JSON object shape with optional
 * fields, and a polymorphic encoding would have to reproduce that shape anyway. The
 * companion factories are the only intended way to build one, so a `location` event cannot
 * pick up a `fromBucket` by accident.
 */
@Serializable
data class TraceEvent(
    val type: TraceEventType,
    /**
     * When the event **happened**, not when this process heard about it.
     *
     * Activity Transition delivery on a Galaxy S21+ was measured at 8.772 s behind the
     * event (docs/04_ANDROID_IMPLEMENTATION.md §20). Recording the receipt time instead
     * would shift every motion event in the trace by the OEM's delivery lag, and the
     * fixture converted from it would describe a drive that never happened. The delivery
     * lag itself belongs in the diagnostics report, which already reports it.
     */
    val atMillis: Long,
    val confidence: TraceConfidence? = null,
    val accuracy: Float? = null,
    val speed: Float? = null,
    val distanceFromPreviousM: Double? = null,
    val fromBucket: LocationQualityBucket? = null,
    val toBucket: LocationQualityBucket? = null,
) {
    companion object {

        fun motion(kind: MotionEventKind, atMillis: Long, confidence: TraceConfidence? = null): TraceEvent =
            TraceEvent(type = kind.traceEventType(), atMillis = atMillis, confidence = confidence)

        fun location(
            atMillis: Long,
            accuracyM: Float,
            speedMps: Float?,
            distanceFromPreviousM: Double?,
        ): TraceEvent = TraceEvent(
            type = TraceEventType.LOCATION,
            atMillis = atMillis,
            accuracy = accuracyM,
            speed = speedMps,
            distanceFromPreviousM = distanceFromPreviousM,
        )

        fun qualityDegraded(
            atMillis: Long,
            fromBucket: LocationQualityBucket,
            toBucket: LocationQualityBucket,
        ): TraceEvent = TraceEvent(
            type = TraceEventType.LOCATION_QUALITY_DEGRADED,
            atMillis = atMillis,
            fromBucket = fromBucket,
            toBucket = toBucket,
        )

        /**
         * The adapter boundary between Android's internal event identities and the
         * cross-platform wire vocabulary.
         *
         * [MotionEventKind.wire] stays `still_enter`/`still_exit` — it is this platform's
         * own naming and changing it would churn the checkpoint, the event log, and the
         * diagnostics report for nothing. The contract vocabulary is
         * `stationary_enter`/`stationary_exit`, and the mapping happens here so the two
         * representations can diverge without either one drifting.
         */
        private fun MotionEventKind.traceEventType(): TraceEventType = when (this) {
            MotionEventKind.ENTERED_VEHICLE -> TraceEventType.VEHICLE_ENTER
            MotionEventKind.EXITED_VEHICLE -> TraceEventType.VEHICLE_EXIT
            MotionEventKind.STARTED_WALKING -> TraceEventType.WALKING_ENTER
            MotionEventKind.BECAME_STATIONARY -> TraceEventType.STATIONARY_ENTER
            MotionEventKind.STOPPED_BEING_STATIONARY -> TraceEventType.STATIONARY_EXIT
        }
    }
}

/**
 * The trace event vocabulary, shared verbatim with iOS and with the fixture converter.
 *
 * The `@SerialName` values are the contract — they are what the converter reads and what
 * the §8 fixtures are written in. Renaming one silently makes the two platforms' traces
 * unusable against each other, which is why they are spelled out here rather than derived
 * from an enum name.
 *
 * Only the subset this platform can currently observe is listed. Car projection and user
 * confirmation events are part of the same vocabulary but belong to features that do not
 * exist yet (M3), and an event type nothing can emit is dead code.
 */
@Serializable
enum class TraceEventType {
    @SerialName("vehicle_enter")
    VEHICLE_ENTER,

    @SerialName("vehicle_exit")
    VEHICLE_EXIT,

    @SerialName("walking_enter")
    WALKING_ENTER,

    @SerialName("stationary_enter")
    STATIONARY_ENTER,

    @SerialName("stationary_exit")
    STATIONARY_EXIT,

    @SerialName("location")
    LOCATION,

    @SerialName("location_quality_degraded")
    LOCATION_QUALITY_DEGRADED,

    ;

    /**
     * The contract string for this type, read back from the `@SerialName` that *is* the
     * contract.
     *
     * The labelling screen shows it so a person cutting a session sees the same vocabulary
     * the fixture will be written in. Derived rather than repeated, because a second
     * hand-written table is a second thing to get out of step.
     */
    val wire: String get() = WIRE_NAMES[ordinal]
}

/** Computed once: [TraceEventType.wire] is read per event while the split list is open. */
private val WIRE_NAMES: List<String> = TraceEventType.entries.map { entry ->
    TraceEventType.serializer().descriptor.getElementName(entry.ordinal)
}

/**
 * docs/05_CROSS_PLATFORM_DOMAIN_CONTRACT.md §5's confidence bucket.
 *
 * Always absent on Android today: `ActivityTransitionEvent` reports a transition without a
 * confidence, unlike iOS `CMMotionActivity`. The field exists so a trace recorded on
 * either platform decodes with the same reader, which is the entire reason §9 is one
 * schema and not two.
 */
@Serializable
enum class TraceConfidence {
    @SerialName("low")
    LOW,

    @SerialName("medium")
    MEDIUM,

    @SerialName("high")
    HIGH,
}

/**
 * Coarse accuracy bands for `location_quality_degraded`
 * (docs/05_CROSS_PLATFORM_DOMAIN_CONTRACT.md §2 `fromBucket`/`toBucket`).
 *
 * **The thresholds are deliberately their own constants and must stay that way.**
 * [kr.parkingpin.app.domain.location.ReliableLocationSelector.MAX_ACCURACY_METERS] happens
 * to be 35 m today, but it is a remote-config-tunable detector threshold
 * (docs/03_SYSTEM_ARCHITECTURE.md §10 `detector.reliableAccuracyMeters`). Deriving the
 * bucket edge from it would mean that tuning the detector retroactively changes what an
 * already-recorded trace says — and a recording format has to stay comparable over time.
 * The coincidence is not a dependency.
 *
 * Both values are field-tuning starting points, in the same spirit as the evidence weights
 * in docs/05_PARKING_DETECTION_ENGINE.md §8.
 */
@Serializable
enum class LocationQualityBucket {
    @SerialName("good")
    GOOD,

    @SerialName("fair")
    FAIR,

    @SerialName("poor")
    POOR,

    ;

    /**
     * Declaration order is the quality order, and [isWorseThan] is the only place that
     * relies on it — a `location_quality_degraded` event is emitted on a drop, never on a
     * recovery, because §2 names the event for the direction it reports.
     */
    fun isWorseThan(other: LocationQualityBucket): Boolean = ordinal > other.ordinal

    companion object {
        const val GOOD_MAX_ACCURACY_METERS: Float = 20f
        const val FAIR_MAX_ACCURACY_METERS: Float = 35f

        /**
         * @return null when [accuracyM] is not a valid accuracy at all.
         *
         * A negative accuracy means Fused Location is telling us the fix is invalid
         * (docs/05_PARKING_DETECTION_ENGINE.md §5), and it is already filtered at the
         * adapter boundary. Giving it a bucket would let an adapter defect hide inside a
         * plausible-looking `poor`, so it gets no bucket and no degradation event; the raw
         * `location` event still carries the negative accuracy, where it is visible.
         */
        fun of(accuracyM: Float): LocationQualityBucket? = when {
            accuracyM < 0f -> null
            accuracyM <= GOOD_MAX_ACCURACY_METERS -> GOOD
            accuracyM <= FAIR_MAX_ACCURACY_METERS -> FAIR
            else -> POOR
        }
    }
}
